### ISSUE TITLE
[Performance] optimizer.step() is extremely slow on AMD GPU, indicating a lack of fused optimizer kernels

### ISSUE THREAD

**Issue by 69gg** [artyom-beilis/pytorch_dlprim#102]
Hi, thank you for your incredible work on this project! I've been working to get a standard GPT-style training script running on my AMD GPU and have successfully navigated several compatibility hurdles thanks to the potential of `pytorch_ocl`.

However, I've hit a definitive performance wall specifically during the `optimizer.step()` phase, which currently makes training impractical. I've conducted extensive debugging and believe I've isolated the root cause.

### High-Level Summary (TL;DR)

The model's forward and backward passes (`model()` and `loss.backward()`) now run efficiently on the GPU after some manual fixes. The entire performance bottleneck (~150 seconds per step) is consumed by `optimizer.step()`. This issue persists identically with both `torch.optim.AdamW` and `torch.optim.SGD`, strongly suggesting a lack of a fused/efficient GPU kernel for the element-wise optimizer update operations. The GPU is effectively idle during this phase.

### Environment

*   **Operating System:** Windows 11
*   **GPU:** AMD Radeon RX 6750 GRE 10GB
*   **GPU Driver Version:** 25.6.3
*   **Python Version:** 3.11.12 (via uv)
*   **PyTorch Version:** 2.6.0+cpu
*   **pytorch_ocl Version:** pytorch_ocl-0.2.0+torch2.6-cp311-none-win_amd64.whl

### Evidence

**1. GPU Utilization Graph:**

This graph is the most critical piece of evidence. It clearly shows two phases within a single training step:
*   **Phase 1 (Left Side):** High, sustained GPU usage. This corresponds to the `model()` forward pass and `loss.backward()` pass. This part is fast and efficient.
*   **Phase 2 (Right Side):** A sharp drop in utilization, followed by low, chaotic, spiky usage. This corresponds to the `optimizer.step()` call and takes over 150 seconds.

<img width="292" height="79" alt="Image" src="https://github.com/user-attachments/assets/68af5b69-e509-405c-af6b-5b158e3391dd" />

**2. Console Log of a Full Iteration:**

This log confirms that a full step completes, but with an unacceptable duration.

```
--- start training---
H:\poetry\.venv\Lib\site-packages\torch\nn\functional.py:2551: UserWarning: The operator 'aten::index_select' is not currently supported on the ocl backend. Please open an issue at for requesting support https://github.com/artyom-beilis/pytorch_dlprim/issues (Triggered internally at C:\Users\artik\Projects\build_env\pytorch_dlprim\src\tensor_ops.cpp:417.)
  return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
step 4600: train loss 1.6312, val loss 2.0709
H:\poetry\.venv\Lib\site-packages\torch\autograd\graph.py:823: UserWarning: The operator 'aten::embedding_dense_backward' is not currently supported on the ocl backend. Please open an issue at for requesting support https://github.com/artyom-beilis/pytorch_dlprim/issues (Triggered internally at C:\Users\artik\Projects\build_env\pytorch_dlprim\src\tensor_ops.cpp:417.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
iter 4600: loss 1.6948, time 154479.71ms, MFU -100.00% (LR: 2.96e-04)
```

### Analysis & Debugging Steps Taken

To isolate the issue, I performed the following steps:

1.  **Fixed Model-Level Bottlenecks:** Initially, `torch.nn.functional.scaled_dot_product_attention` and `masked_fill` were causing CPU fallbacks. I successfully replaced them with manual, GPU-friendly implementations. This is why the forward/backward pass (Phase 1) is now fast.

2.  **Disabled Incompatible Features:** I have disabled AMP, `torch.compile`, and `pin_memory`, as they are not currently supported by the OCL backend.

3.  **Tested Multiple Optimizers:** The key experiment was switching from `torch.optim.AdamW` to `torch.optim.SGD`. The performance bottleneck remained identical (~150s per step), proving the issue is not with a specific algorithm's complexity, but with the fundamental way the `step()` function is dispatched for `ocl` devices.

4.  **Verified No Special Optimizer Class:** My investigation confirmed that this library uses backend overrides rather than providing a separate optimizer class like `pytorch_ocl.optim.AdamW`. The standard `torch.optim` classes are intended to be used.

### Conclusion and Request

The evidence overwhelmingly points to the `optimizer.step()` implementation for the `ocl` backend lacking efficient, fused GPU kernels. The current approach seems to cause extreme overhead, likely from dispatching millions of tiny, element-wise computations individually, which negates all the performance gains from the GPU-accelerated model computation.

This currently makes `pytorch_ocl` a powerful tool for inference, but challenging to use for practical training.

*   Could you please confirm if this is a known limitation for training workloads?
*   Are there any plans in the roadmap to implement fused kernels for common optimizers like AdamW and SGD?

Thank you again for this fantastic project. I am very willing to help test any potential fixes or new versions.

(sorry for bad English)

**Comment by artyom-beilis** [comment]
First of all very good analysis but the clue is there

```
UserWarning: The operator 'aten::index_select' is not currently supported on the ocl backend. 
UserWarning: The operator 'aten::embedding_dense_backward' is not currently supported on the ocl backend.
```

That means it fall-backs to CPU since it isn't implemented in GPU yet. It isn't about "merged" kernels it is far worse `:-)`

> Could you please confirm if this is a known limitation for training workloads?

Yes there are still many operators are missing and some work loads (particularly LLMs and some Yolo like stuff) misses some essential operators.

> Are there any plans in the roadmap to implement fused kernels for common optimizers like AdamW and SGD?

Fused kernels for AdamW/SDG currently is less critical problem (I'm aware of) and gives only very small performance cost

The real problem is lack of several embedding and indexing related operators.

And I do want to work on it - not sure exactly when - mostly because it depends on availability of my free time I have to spend on this project.

Anyway I want to start digging into LLMs and NLP so probably I'll start working on it.

**Comment by 69gg** [comment]
Thank you for the incredibly fast and clear response!
Your explanation makes perfect sense. The problem being the widespread CPU fallbacks of fundamental operators (like embedding/indexing) rather than just the optimizer itself clarifies everything. I really appreciate you taking the time to explain the technical details.
It's fantastic news to hear that you're interested in digging into LLMs and NLP workloads in the future. My project and I would be more than happy to serve as a test case for any potential updates in this area.
Thanks again for your hard work on this amazing project and for clarifying the situation. I'll be eagerly watching for future updates!
(sorry for bad English again and again :) )
